cstore_reader.c

```diff
@@ -200,6 +214,85 @@ CStoreReadFooter(StringInfo tableFooterFilename)
 }
 
 
+/* DeserializeBlockData  deserializes requested data block for all columns. It
+ * uncompresses serialized data if necessary, sets deserialized buffer index at
+ * read state. The function also deallocates previous data buffer used to keep temporary
+ * uncompressed data buffer.
+ */
+static void
+DeserializeBlockData(TupleDesc tupleDescriptor, StripeData *stripeData,
+					 MemoryContext stripeReadContext, uint64 blockIndex,
+					 BlockDataBuffers **blockDataBuffersArray)
+{
+	int columnIndex = 0;
+	MemoryContext oldContext = MemoryContextSwitchTo(stripeReadContext);
+
+	for (columnIndex=0; columnIndex < stripeData->columnCount; columnIndex++)
+	{
+		ColumnData *columnData = stripeData->columnDataArray[columnIndex];
+
+		if (columnData != NULL)
+		{
+			ColumnBlockData *blockData = columnData->blockDataArray[blockIndex];
+			Form_pg_attribute *attributeFormArray = tupleDescriptor->attrs;
+			bool typeByValue = attributeFormArray[columnIndex]->attbyval;
+			int typeLength = attributeFormArray[columnIndex]->attlen;
+			char typeAlign = attributeFormArray[columnIndex]->attalign;
+			BlockDataBuffers * blockDataBuffers = blockDataBuffersArray[columnIndex];
+			StringInfo valueBuffer = NULL;
+
+			if (blockDataBuffers->serializedValueBuffer != NULL)
+			{
+				pfree(blockDataBuffers->serializedValueBuffer->data);
+				pfree(blockDataBuffers->serializedValueBuffer);
+				blockDataBuffers->serializedValueBuffer = NULL;
+			}
+
+			valueBuffer = DecompressBuffer(blockData->valueBuffer,
+										   blockData->valueCompressionType);
+
+			DeserializeBoolArray(blockData->existBuffer,
+								 blockDataBuffers->existsArray,
+								 blockData->rowCount);
+			DeserializeDatumArray(valueBuffer, blockDataBuffers->existsArray,
+								  blockData->rowCount, typeByValue, typeLength,
+								  typeAlign, blockDataBuffers->valueArray);
+
+			/* store newly allocated uncompressed data buffer to be freed later */
+			if (valueBuffer != blockData->valueBuffer)
+			{
+				blockDataBuffers->serializedValueBuffer = valueBuffer;
+			}
+		}
+	}
+
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+/*
+ * ResetUncompressedDataBuffers iterates over deserialized column block data
+ * buffers and sets uncompressedDataBuffer field to NULL. This field is
+ * allocated in stripe memory context and becomes invalid once memory context
+ * is reset.
+ */
+static void
```

Comment:
Here the comment says "... and sets uncompressedDataBuffer field to NULL" while we are actually setting ```serializedValueBuffer``` to ```NULL``` and not ```uncompressedDataBuffer```. Other details of the comment should be double checked, just to make sure.